x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java

```diff
@@ -0,0 +1,163 @@
+            return roleNames;
+        }
+
+        @Override
+        public RoleKey cacheKey() {
+            if (roleNames.length == 0) {
+                return RoleKey.ROLE_KEY_EMPTY;
+            } else if (Arrays.asList(roleNames).contains(ReservedRolesStore.SUPERUSER_ROLE_DESCRIPTOR.getName())) {
+                return RoleKey.ROLE_KEY_SUPERUSER;
+            } else {
+                return new RoleKey(Set.copyOf(new HashSet<>(List.of(roleNames))), RoleKey.ROLES_STORE_SOURCE);
+            }
+        }
+
+        @Override
+        public void resolve(RoleReferenceResolver resolver, ActionListener<RolesRetrievalResult> listener) {
+            resolver.resolveNamedRoleReference(this, listener);
+        }
+    }
+
+    /**
+     * Referencing API Key role descriptors. Can be either the assigned (key) role descriptors or the limited-by (owner's) role descriptors
+     */
+    final class ApiKeyRoleReference implements RoleReference {
+
+        private final String apiKeyId;
+        private final BytesReference roleDescriptorsBytes;
+        private final String roleKeySource;
+
+        public ApiKeyRoleReference(String apiKeyId, BytesReference roleDescriptorsBytes, String roleKeySource) {
+            this.apiKeyId = apiKeyId;
+            this.roleDescriptorsBytes = roleDescriptorsBytes;
+            this.roleKeySource = roleKeySource;
+        }
+
+        @Override
+        public RoleKey cacheKey() {
+            final String roleDescriptorsHash = MessageDigests.toHexString(
+                MessageDigests.digest(roleDescriptorsBytes, MessageDigests.sha256())
+            );
+            return new RoleKey(Set.of("apikey:" + roleDescriptorsHash), roleKeySource);
+        }
+
+        @Override
+        public void resolve(RoleReferenceResolver resolver, ActionListener<RolesRetrievalResult> listener) {
+            resolver.resolveApiKeyRoleReference(this, listener);
+        }
+
+        public String getApiKeyId() {
+            return apiKeyId;
+        }
+
+        public BytesReference getRoleDescriptorsBytes() {
+            return roleDescriptorsBytes;
+        }
+    }
+
+    /**
+     * Same as {@link ApiKeyRoleReference} but for BWC purpose (prior to v7.9.0)
+     */
+    final class BwcApiKeyRoleReference implements RoleReference {
+        private final String apiKeyId;
+        private final Map<String, Object> roleDescriptorsMap;
+        private final String roleKeySourceSuffix;
+
+        public BwcApiKeyRoleReference(String apiKeyId, Map<String, Object> roleDescriptorsMap, String roleKeySourceSuffix) {
+            this.apiKeyId = apiKeyId;
+            this.roleDescriptorsMap = roleDescriptorsMap;
+            this.roleKeySourceSuffix = roleKeySourceSuffix;
+        }
+
+        @Override
+        public RoleKey cacheKey() {
+            // Since api key id is unique, it is sufficient and more correct to use it as the names
+            return new RoleKey(Set.of(apiKeyId), apiKeyId + roleKeySourceSuffix);
```

Comment:
The existing code use API key role names as the first argument to `RoleKey' constructor. It's hard to get that information with the new structure and it is also unnecessary. Since API key ID is used as the source of `RoleKey`. It is already guaranteed that one entry for each single key. So I just use the API key ID again as the first argument.

Reply:
Why do we include the api key id in the source field? Is that needed?

Reply:
It is not needed. I just copied it from the existing code. But it was needed there because the names are not guaranteed to be unique. Since we now use the API Key ID as the names, it is no longer necesesary to have it in the source. I'll drop it.